float64[] z1OL
float64[] z2OL
float64[] z3OL
float64[] z4OL
float64[] u1OL
float64[] u2OL